### PR TITLE
`Communication`: Change the Markdown editor to Monaco

### PR DESCRIPTION
**Pull request by pzdr7**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

We want to replace the Ace editor with Monaco in Artemis. One important use case of Ace currently is the Communication view, where the markdown editor is used to
- mention users
- reference channels
- link lectures, slides, attachments and attachment units

There are a few usability issues with the current implementation. In particular, the suggest widget currently used by the markdown editor lacks keyboard input and interferes with the layout of the editor around it.
<details>
  <summary>Layout problem examples</summary>

image

image

</details>

Some users also experience lag when holding down a button and typing in the editor.

### Description

- Changed a few components' change detection strategy to `OnPush`, reducing the lag on the communication page.
- Replaced the Ace editor on the Communication page with Monaco.
  - User mentions and channel references trigger the built-in Monaco suggest (via \@ and \#).
  - Exercise references now work via the suggest widget as well
  - Lecture/Slide/Attachment/Attachment Unit references still work the same way as before - via the editor bar.
  - The editor grows with the content.
  - No longer inserts a newline when hitting enter to send

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Course with 
  - Communication and messaging enabled
  - At least one lecture that has one attachment and at least one unit with slides. You can set up such a lecture by using the "Automatic unit processing" button when you create it. [Some slides you can use](https://github.com/user-attachments/files/16397447/Infun2324.L02.Control.Structures.1.pdf)

> [!IMPORTANT]  
> For a faster setup, you can also use [my test course on TS1](https://artemis-test1.artemis.cit.tum.de/courses/18/communication). L1/L2 have either attachments or units you can reference

1. Log in to Artemis
2. Navigate to the course communications
3. Write some messages. In particular, try:
- Writing a message in a channel, a group chat, and a one-on-one chat. You can hit Enter to send
- Replying to a message in a thread
- Editing a message
- Making an announcement
- Using \# (or the corresponding action in the editor's bar) to reference a channel
- Using \@ (or the corresponding action in the editor's bar) to mention a user
- Using /exercise (or the corresponding action in the editor's bar) to reference an exercise
- Referencing a lecture, an attachment, an attachment unit, or a slide via the action in the editor bar (this is largely unchanged due to the complexity of the data structure here) **Note the known issues.**

**Known issues**:

- #7980 Linking lectures and exercises hasn't been possible in the "Communication" of an exercise or lecture for a while. This will be fully fixed in a follow-up.
- Some test servers seem to have a bug where the server doesn't find all users of a course, e.g. TS3.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| markdown-editor-monaco.component.ts | 98.41% | ✅ |
| answer-post.component.ts | 100% | ✅ |
| posting-markdown-editor.component.ts | 100% | ✅ |
| monaco-channel-reference.action.ts | 100% | ✅ |
| monaco-exercise-reference.action.ts | 100% | ✅ |
| monaco-lecture-attachment-reference.action.ts | 100% | ✅ |
| monaco-user-mention.action.ts | 100% | ✅ |
| monaco-editor-action.model.ts | 95.58% | ✅ |
| monaco-editor.component.ts | 91.12% | ✅ |

### Screenshots

image

image

image

## Summary by CodeRabbit

- **New Features**
	- Enhanced user interactivity in conversation threads with improved scrolling and event handling.
	- Introduced new input properties and output event emitters for better functionality in the Markdown editor.
	- Added support for inserting references in the Monaco editor with new action classes, including user mentions and channel references.
	- Enabled users to create new lines in the markdown editor by pressing Enter.
	- Transitioned to a Monaco editor, improving user experience with advanced features.

- **Bug Fixes**
	- Improved change detection and performance across various components.

- **Tests**
	- Introduced new test cases for the Markdown editor and refactored existing tests for better structure.

### CODE REVIEW

File: src/main/webapp/app/shared/metis/message/message-reply-inline-input/message-reply-inline-input.component.html [hunk 1]
```diff
@@ -1,13 +1,14 @@
 <form [formGroup]="formGroup" (ngSubmit)="confirm()">
-    <div class="message-inline-input-form__wrapper mb-3">
+    <div class="message-inline-input-form__wrapper mb-3 me-1">
         <jhi-posting-markdown-editor
             formControlName="content"
             [editorHeight]="editorHeight"
             [maxContentLength]="maxContentLength"
             [isInputLengthDisplayed]="false"
             (keydown.enter)="confirm()"
+            (valueChange)="valueChange.emit()"
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/shared/metis/message/message-reply-inline-input/message-reply-inline-input.component.html:9]: It is handled in conversation-thread-sidebar.component.html

File: src/main/webapp/app/shared/monaco-editor/model/actions/communication/monaco-channel-reference.action.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        this.typeText(editor, MonacoChannelReferenceAction.DEFAULT_INSERT_TEXT);
+        editor.focus();
+    }
+
+    async fetchChannels(): Promise<ChannelIdAndNameDTO[]> {
+        if (!this.cachedChannels) {
+            const response = await firstValueFrom(this.channelService.getPublicChannelsOfCourse(this.metisService.getCourse().id!));
+            this.cachedChannels = response.body!;
+        }
+        return this.cachedChannels;
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/shared/monaco-editor/model/actions/communication/monaco-channel-reference.action.ts:62]: This action will only be used in a context where the course ID is available

File: src/main/webapp/app/shared/monaco-editor/model/actions/communication/monaco-lecture-attachment-reference.action.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                    .filter((lecture) => !!lecture.id && !!lecture.title)
+                    .map((lecture) => ({
+                        id: lecture.id!,
+                        title: lecture.title!,
+                        attachmentUnits: lecture.lectureUnits?.filter((unit) => unit.type === LectureUnitType.ATTACHMENT),
+                        attachments: lecture.attachments,
+                    }));
+            }
+        });
+    }
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/shared/monaco-editor/model/actions/communication/monaco-lecture-attachment-reference.action.ts:54]: This action will only be used in a context where the course ID is available

File: src/main/webapp/app/shared/monaco-editor/model/actions/communication/monaco-user-mention.action.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    dispose(): void {
+        super.dispose();
+        this.disposableCompletionProvider?.dispose();
+    }
+
+    async loadUsersForSearchTerm(searchTerm: string): Promise<UserNameAndLoginDTO[]> {
+        const response = await firstValueFrom(this.courseManagementService.searchMembersForUserMentions(this.metisService.getCourse().id!, searchTerm));
+        return response.body ?? [];
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/shared/monaco-editor/model/actions/communication/monaco-user-mention.action.ts:65]: This action will only be used in a context where the course ID is available

File: src/test/javascript/spec/component/shared/monaco-editor/monaco-editor-communication-action.integration.spec.ts [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
+
+    it('should create', () => {
+        fixture.detectChanges();
+        expect(comp).toBeTruthy();
+    });
+
```
Comment (pzdr7) [inline src/test/javascript/spec/component/shared/monaco-editor/monaco-editor-communication-action.integration.spec.ts:82]:
```suggestion
```

File: src/test/javascript/spec/component/shared/monaco-editor/monaco-editor-communication-action.integration.spec.ts [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+        let exercises: Exercise[];
+
+        beforeEach(() => {
+            fixture.detectChanges();
+            comp.changeModel('initial', '');
+            channels = [metisGeneralChannelDTO, metisExamChannelDTO, metisExerciseChannelDTO];
+            channelReferenceAction.cachedChannels = channels;
+            users = [metisUser1, metisUser2, metisTutor];
+            jest.spyOn(courseManagementService, 'searchMembersForUserMentions').mockReturnValue(of(new HttpResponse({ body: users, status: 200 })));
+            exercises = metisService.getCourse().exercises!;
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/test/javascript/spec/component/shared/monaco-editor/monaco-editor-communication-action.integration.spec.ts:112]: This is fine in the tests

**Review by egekurt123 — DISMISSED** [review]
Tested on TS1, Works as expected

**Review by Jan-Thurner — DISMISSED** [review]
tested on TS1, works like a charm.

**Review by JohannesWt — COMMENTED** [review]
Code looks good just one comment

File: src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html [hunk 7]
```diff
@@ -98,6 +105,82 @@
...
+                                <span [jhiTranslate]="displayedActions.lecture.translationKey"></span>
+                                <fa-icon [icon]="faAngleDown" class="ms-1" />
+                            </button>
+                            <mat-menu #subMenuLecture="matMenu" type="button">
+                                @for (lecture of displayedActions.lecture.lecturesWithDetails; track lecture.id) {
+                                    <button class="ps-1 me-1" mat-menu-item [matMenuTriggerFor]="lectureMenuUnits" type="button">
+                                        {{ lecture.title }}
+                                    </button>
+                                    <mat-menu #lectureMenuUnits="matMenu" type="button">
+                                        <button
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html:120]: I'm not entirely sure if this is possible within a mat-menu, as some mat-components require a specific tag structure. However, you could try adding an ng-template for the buttons.

**Review by rabeatwork — CHANGES_REQUESTED** [review]

**Review by BBesrour — APPROVED** [review]
Tested on TS1, works as described

File: src/main/webapp/app/shared/metis/posting-markdown-editor/posting-markdown-editor.component.html [hunk 8]
```diff
@@ -1,30 +1,34 @@
...
         [markdown]="content"
         (markdownChange)="updateField($event)"
-        [resizableMinHeight]="editorHeight"
         (onEditSelect)="previewMode = false"
         (onPreviewSelect)="previewMode = true"
-        [metaCommands]="[]"
-        [headerCommands]="[]"
-        [defaultCommands]="defaultCommands"
+        [metaActions]="[]"
+        [headerActions]="undefined"
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/shared/metis/posting-markdown-editor/posting-markdown-editor.component.html:9]: These currently overwrite some default actions of the Markdown editor. Removing `undefined` would enable those actions.

I'll have to see once all Markdown editors have been replaced if this makes sense to keep (or if we should do the opposite & have undefined be the default).

File: src/main/webapp/app/shared/monaco-editor/model/actions/monaco-editor-action.model.ts [hunk 9]
```diff
@@ -79,6 +79,106 @@ export abstract class MonacoEditorAction implements monaco.editor.IActionDescrip
...
+        if (triggerCharacter !== undefined && triggerCharacter.length !== 1) {
+            throw new Error(`The trigger character must be a single character.`);
+        }
+        const languageId = model.getLanguageId();
+        const modelId = model.id;
+        // We have to subtract an offset of 1 from the start column to include the trigger character in the range that will be replaced.
+        const triggerCharacterOffset = triggerCharacter ? 1 : 0;
+        return monaco.languages.registerCompletionItemProvider(languageId, {
+            // We only want to trigger the completion provider if the trigger character is typed. However, we also allow numbers to trigger the completion, as they would not normally trigger it.
+            triggerCharacters: triggerCharacter ? [triggerCharacter, '1', '2', '3', '4', '5', '6', '7', '8', '9', '0'] : undefined,
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/shared/monaco-editor/model/actions/monaco-editor-action.model.ts:110]: 👍

File: src/main/webapp/app/shared/monaco-editor/model/actions/monaco-editor-action.model.ts [hunk 10]
```diff
@@ -79,6 +79,106 @@ export abstract class MonacoEditorAction implements monaco.editor.IActionDescrip
...
+                });
+
+                // We only want suggestions if the trigger character is at the beginning of the word.
+                if (triggerCharacter) {
+                    if (sequenceUntilPosition.word !== triggerCharacter) {
+                        if (beforeWord !== triggerCharacter) {
+                            return undefined;
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/shared/monaco-editor/model/actions/monaco-editor-action.model.ts:139]: Forgot to change this after debugging 😅

File: src/main/webapp/app/shared/monaco-editor/model/actions/monaco-editor-action.model.ts [hunk 11]
```diff
@@ -79,6 +79,106 @@ export abstract class MonacoEditorAction implements monaco.editor.IActionDescrip
...
+                return searchFn(sequenceUntilPosition.word).then((items) => {
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/shared/monaco-editor/model/actions/monaco-editor-action.model.ts:140]: Good idea 👍

File: src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html (hunk 7 above)
Comment (pzdr7) [reply src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html:120]: I tried this - unfortunately, it does break the menu. Specifically, the hover trigger to open a sub-menu no longer works. The sub-menu still opens on click, but the original menu closes, so the sub-menu is just floating around on the page.

**Review by Jan-Thurner — APPROVED** [review]
tested again on TS1. still works well.

**Review by BBesrour — APPROVED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by rabeatwork — APPROVED** [review]
Thanks for incorporating my feedback.

**Review by egekurt123 — APPROVED** [review]
Reapprove

**Review by laadvo — APPROVED** [review]
Tested on ts1, the editor is working as expected

**Review by JohannesWt — APPROVED** [review]
Code Reapprove

**Pull request merged**